File: leetcode_c++/Main_0590_N叉树的后序遍历.cpp
// main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <unordered_set>

using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};

class Solution {
public:
    vector<int> postorder(Node* root) {
    
        vector<int> res;
        if (root == nullptr) {
            return res;
        }

        stack<Node *> st;
        unordered_set<Node *> visited;
        st.emplace(root);
        while (!st.empty()) {
            Node * node = st.top();
            /* 如果当前节点为叶子节点或者当前节点的子节点已经遍历过 */
            if (node->children.size() == 0 || visited.count(node)) {
                res.emplace_back(node->val);
                st.pop();
                continue;
            }
            for (auto it = node->children.rbegin(); it != node->children.rend(); it++) {
                st.emplace(*it);
            }
            visited.emplace(node);
        }       
        return res;
    }
};

int main()
{

    printf("Hello Main_0590_N叉树的后序遍历");

    return 0;
}

File: leetcode_java/src/main/java/com/ariescat/metis/leetcode/Main_0590_N叉树的后序遍历.java
package com.ariescat.metis.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Main_0590_N叉树的后序遍历 {

    public static void main(String[] args) {
        Node root = new Node(1, new ArrayList<Node>() {
            {
                add(new Node(3, new ArrayList<Node>() {{
                    add(new Node(5));
                    add(new Node(6));
                }}));
                add(new Node(2));
                add(new Node(4));
            }
        });
        List<Integer> postorder = postorder(root);
        System.err.println(postorder);
    }

    public static List<Integer> postorder(Node root) {
        List<Integer> res = new ArrayList<>();
        postorder0(root, res);
        return res;
    }

    public static void postorder0(Node root, List<Integer> res) {
        if (root == null) {
            return;
        }
        for (Node ch : root.children) {
            postorder0(ch, res);
        }
        res.add(root.val);
    }


    static class Node {
        public int val;
        public List<Node> children;

        public Node() {
        }

        public Node(int _val) {
            val = _val;
        }

        public Node(int _val, List<Node> _children) {
            val = _val;
            children = _children;
        }
    }
}
